homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,173 @@
+        [ElecPriceSensor(name, config_entry.unique_id, pvpc_data_handler)], True
+    )
+
+
+class ElecPriceSensor(RestoreEntity):
+    """Class to hold the prices of electricity as a sensor."""
+
+    unit_of_measurement = UNIT
+    icon = ICON
+    should_poll = False
+
+    def __init__(self, name, unique_id, pvpc_data_handler):
+        """Initialize the sensor object."""
+        self._name = name
+        self._unique_id = unique_id
+        self._pvpc_data = pvpc_data_handler
+        self._num_retries = 0
+
+        self._init_done = False
+        self._hourly_tracker = None
+        self._price_tracker = None
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Cancel listeners for sensor updates."""
+        self._hourly_tracker()
+        self._price_tracker()
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._pvpc_data.state = state.state
+
+        # Update 'state' value in hour changes
+        self._hourly_tracker = async_track_time_change(
+            self.hass, self.async_update, second=[0], minute=[0]
+        )
+        # Update prices at random time, 2 times/hour (don't want to upset API)
+        random_minute = randint(1, 29)
+        mins_update = [random_minute, random_minute + 30]
+        self._price_tracker = async_track_time_change(
+            self.hass, self.async_update_prices, second=[0], minute=mins_update,
+        )
+        _LOGGER.debug(
+            "Setup of price sensor %s (%s) with tariff '%s', "
+            "updating prices each hour at %s min",
+            self.name,
+            self.entity_id,
+            self._pvpc_data.tariff,
+            mins_update,
+        )
+        await self.async_update_prices(dt_util.utcnow())
+        self._init_done = True
+        await self.async_update_ha_state(True)
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._pvpc_data.state
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._pvpc_data.state_available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._pvpc_data.attributes
+
+    async def async_update(self, *args):
+        """Update the sensor state."""
+        if not self._init_done:
+            # abort until added_to_hass is finished
+            return
+
+        now = dt_util.utcnow() if not args else args[0]
+        self._pvpc_data.process_state_and_attributes(now)
+        self.async_write_ha_state()
```

Comment:
```suggestion
```

You should be able to remove this as 

`homeassistant.helpers.entity.async_update_ha_state` will call it after the update.

Reply:
> You should be able to remove this `self.async_write_ha_state()`

No. Only the calls from `added_to_hass` or when data source recovers availability are using `await self.async_update_ha_state(True)` or `async_schedule_update_ha_state(True)` to trigger the update.

To make things clearer, I could just call `async_update()` in those two.

The periodic updates (on x:00:00 times) to change the current price are handled with the `self._hourly_tracker`, which calls `async_update`. As `should_poll` is False, those are the only calls, so the state must be written at that moment.

> Can you make this simpler and change the call to async_add_entities in async_setup_entry from True to False so it doesn't do an update before adding?

Yes! I didn't remember what the boolean on `add_entities` was for. Now I don't need the `self._init_done` flag anymore.

Instead of:

```python
await self.async_update_prices(dt_util.utcnow())
self._init_done = True
await self.async_update_ha_state(True)
```

I could do just:

```python
await self.async_update_prices(dt_util.utcnow())
await self.async_update()
```

and the same after `_LOGGER.warning("Component has recovered data access")`, instead of using `self.async_schedule_update_ha_state(True)`

Reply:
If the user calls the generic update service calls this, its going to write state afterwards so you'll get two writes.

Reply:
> If the user calls the generic update service calls this, its going to write state afterwards so you'll get two writes.

I don't follow you, sorry. What do you mean with 'If the user calls the generic update service calls'?

Also, if I understand you (I think not), what you want me to remove is the call to `self.async_write_ha_state()` inside `async_update`, right?
If I remove that, when hour changes, the data handler internally changes the price state, but the HA state machine is not notified and price doesn't change. I need to write_state there.

Reply:
Let me push up an example.  Give me a few minutes.

Reply:
Maybe something like this

https://github.com/bdraco/home-assistant/blob/f22243aea484b00c52ca039507eff8777c30c5f4/homeassistant/components/pvpc_hourly_pricing/sensor.py#L134

Reply:
You are confusing how the integration works, you cannot remove the `_hourly_tracker`.

There are '2 update processes':
- `async_update` is called each change of hour, to change the current hourly price. This is not an IO operation, as we have stored 24 prices (for each day). We are only selecting the one corresponding with the current hour.
- `async_update_prices` is the one downloading prices, and it is called 2 times/hour, at random times. It could just be called once a day (not a good idea IMO because many reasons, but it could be done)

Because of that, there are _2 flags of unavailability_:
- `self._pvpc_data.source_available` indicates if any download was successful, and there are prices for today. Unless serious problem or unavailable internet, this will be True always.
- `self._pvpc_data.state_available` is linked to the `entity.available` prop, as indicates if there is a current price (aka state).

Imagine that at 11:23 the internet connection is lost. Subsequent calls to `async_update_prices` will fail, and eventually, the `source_available` flag would turn off.
Without internet connection, **the sensor state will be available until 00:00, as we already have all the prices for 'today'**. At 00:00, there won't be a valid price to set the state, and at that moment the `state_available` flag will turn off.
If, on the contrary, the network is recovered at any time on the same day, the sensor would not lose the availability at any moment.

Also, we **don't want** to call the API to download prices when updating state, as we need that update to be as fast as possible. Why? because making a request is always going to take some time, and then you'll have a price change with a timestamp far from exact (there will be obviously some delay always, but milliseconds matter, for example, when integrating with `utility_meter`s).

**Maybe it would be clearer if I rename the `async_update` to something else**, and make it a `callback` (as it is safe to call from within the event loop), as there is no real need to have an `update` method in the entity when `should_poll=False`.

In that case, `added_to_hass` would include:

```python
# Update 'state' value in hour changes
self._hourly_tracker = async_track_time_change(
    self.hass, self.update_current_price, second=[0], minute=[0]
)

# ...

now = dt_util.utcnow()
await self.async_update_prices(now)
self.update_current_price(now)
```

and the method would be:

```python
@callback
def update_current_price(self, now):
    """Update the sensor state."""
    self._pvpc_data.process_state_and_attributes(now)
    self.async_write_ha_state()
```

**That change works as expected, so I can push it if you consider it better than the current one**, but the change you proposed breaks the sensor behavior and it's not optimal because of the explanation above, I hope you understand.

Reply:
Thanks for the detailed explanation.  Apologizes for missing this detail.  I started off with the bad assumption that the pypi package was handling data freshness.  I’ll run though this again after breakfast

Reply:
@azogue  As it stands how it looks like calling `async_update` manually via `homeassistant.update_entity` (https://www.home-assistant.io/docs/scripts/service-calls/) won't ever change the state of the sensor as there will never be new data on a manual call since its always called right after updating the price data.  

The change you are proposing makes it seem much cleaner to me :)

Reply:
> @azogue As it stands how it looks like calling async_update manually via homeassistant.update_entity

Wow, I'm totally outdated :), I didn't know about `homeassistant.update_entity`!
Well, as that service ends calling the `async_update` method, calling it would be kind of dumb, as it would do 2 `self.async_write_ha_state()` calls, but no `EVENT_STATE_CHANGED` would be fired, as state won't change. But yes, better to rename it to something different than `async_update`. I'll push it in a bit